web/src/inbox_ui.ts

```diff
@@ -821,6 +903,83 @@ function reset_data(): {
+                other_channels_folder.is_header_visible =
+                    other_channels_folder.is_header_visible || !stream_data.is_hidden;
+                other_channels_folder.has_unread_mention =
+                    other_channels_folder.has_unread_mention || stream_data.mention_in_unread;
+            }
+        } else if (stream_data.folder_id === PINNED_CHANNEL_FOLDER_ID) {
+            const pinned_channel_folder = channel_folders_dict.get(PINNED_CHANNEL_FOLDER_ID);
+            if (pinned_channel_folder === undefined) {
+                channel_folders_dict.set(PINNED_CHANNEL_FOLDER_ID, {
+                    id: PINNED_CHANNEL_FOLDER_ID,
+                    header_id: PINNED_CHANNEL_HEADER_ID,
+                    name: $t({defaultMessage: "PINNED CHANNELS"}),
+                    is_header_visible: !stream_data.is_hidden,
+                    unread_count: stream_data.unread_count,
+                    is_collapsed: collapsed_containers.has(PINNED_CHANNEL_HEADER_ID),
+                    has_unread_mention: stream_data.mention_in_unread,
+                });
+            } else {
+                pinned_channel_folder.unread_count =
+                    (pinned_channel_folder.unread_count ?? 0) + (stream_data.unread_count ?? 0);
+                pinned_channel_folder.is_header_visible =
+                    pinned_channel_folder.is_header_visible || !stream_data.is_hidden;
+                pinned_channel_folder.has_unread_mention =
+                    pinned_channel_folder.has_unread_mention || stream_data.mention_in_unread;
+            }
+        } else {
+            const folder_id = stream_data.folder_id;
+            let channel_folder = channel_folders_dict.get(folder_id);
+            if (channel_folder === undefined) {
+                channel_folder = {
+                    id: folder_id,
+                    header_id: CHANNEL_FOLDER_HEADER_ID_PREFIX + folder_id,
+                    name: channel_folders.get_channel_folder_name_from_id(folder_id),
+                    is_header_visible: !stream_data.is_hidden,
+                    unread_count: stream_data.unread_count,
+                    is_collapsed: collapsed_containers.has(
+                        CHANNEL_FOLDER_HEADER_ID_PREFIX + folder_id,
+                    ),
+                    has_unread_mention: stream_data.mention_in_unread,
+                };
+                channel_folders_dict.set(folder_id, channel_folder);
+            } else {
+                channel_folder.unread_count =
+                    (channel_folder.unread_count ?? 0) + (stream_data.unread_count ?? 0);
+                channel_folder.is_header_visible =
+                    channel_folder.is_header_visible || !stream_data.is_hidden;
+                channel_folder.has_unread_mention =
+                    channel_folder.has_unread_mention || stream_data.mention_in_unread;
+            }
+        }
+    }
```

Comment:
I think you can probably deduplicate this a bunch with a couple functions -- one to make a `folder_context` object (probably a better name than `channel_folder`, since the latter should be reserved for an API object), and a `update_unread_counts(folder_context, channel_context)` for the second part of this repeated pattern.

Reply:
Moved the entire `if/else` clause to a new function.